### PR TITLE
programs: add vdirsyncer

### PR DESCRIPTION
**Pull request by loewenheim**
This adds support for the `vdirsyncer` utility.

### CODE REVIEW

**Comment by dtzWill** [comment]
And just yesterday I fumbled my way through setting up vdirsyncer! haha

Consider supporting these options maybe?
* `password.fetch` (so password isn't globally readable in nix store)
* `metadata` - array of metadata to sync, I have `["displayname", "color"]` for my calendar and `["displayname"]` for contacts-- although I'm still new so mostly setting them since the docs mention it :grin:.

**Comment by loewenheim** [comment]
@dtzWill You're completely right. I actually use both those options myself. I'll add them as soon as I can.

File: modules/programs/vdirsyncer.nix [hunk 1]
```diff
@@ -0,0 +1,404 @@
...
+        default = null;
+        description = ''
+          Configuration for vdirsyncer.
+        '';
+        type = types.nullOr (types.submodule {
+          options = {
+
+            statusPath = mkOption {
+              type = types.str;
+              default = "$HOME/.vdirsyncer/status";
```
Comment (teto) [inline modules/programs/vdirsyncer.nix]: This defies XDG. I would use XDG_DATA_HOME

File: modules/programs/vdirsyncer.nix [hunk 2]
```diff
@@ -0,0 +1,404 @@
...
+  
+  ## Type definitions
+  collection = types.either types.str (types.listOf types.str);
+  storageType = types.enum [
+    "caldav" "carddav" "http"
+    "filesystem" "singlefile"
+  ];
+
+  storage = types.submodule {
+    options = {
```
Comment (teto) [inline modules/programs/vdirsyncer.nix]: what about localStorage vs remoteStorage ? can we really create pairs between 2 caldav for instance ?

Comment (teto) [reply modules/programs/vdirsyncer.nix]: the storage might need an extraConfig option. For instance with the type "goole_calendar", I have to set 
```
# token_file = "~/.local/vdirsyncer"
client_id = "dsds"
client_secret = "~/.config/vdirsyncer/client_secret"

```

Comment (loewenheim) [reply modules/programs/vdirsyncer.nix]: > what about localStorage vs remoteStorage ? can we really create pairs between 2 caldav for instance ?

To be honest, I just assumed that this was possible, but I’m going to test it.

> the storage might need an extraConfig option. For instance with the type "goole_calendar", I have to set

Currently, you can’t create a storage of type "google_calendar" anyway because the allowed types are hardcoded and I haven’t implemented the google and ete stuff yet. It’s on the agenda :)

File: modules/programs/vdirsyncer.nix (hunk 1 above)
Comment (loewenheim) [reply modules/programs/vdirsyncer.nix]: You’re right, I copied the example from the tutorial, but XDG_DATA_HOME makes more sense.

**Comment by teto** [comment]
@loewenheim would you be interested in abstracting that a bit similar to  the email instrastructure ? like creating a modules/accounts/calendars.nix on the model of modules/accounts/email.nix ?
so that we could that then configure once and have khal/vdirsyncer/rhemynd and other calendar program configurations in sync ?

I would review/test :)

**Comment by loewenheim** [comment]
@teto Yes, absolutely. That sounds like a great idea!

**Comment by loewenheim** [comment]
Here’s what I’m thinking, designwise: A `calendar` should subsume both a `pair` and its two `storage`s. We assume that there is always one local and one remote storage. The calendar would have submodules for local (e.g., whether to save it as a directory or a single file) and remote (e.g., authentication stuff) settings.

Does that sound sensible?

**Comment by teto** [comment]
I think it's ok. Programs like khal only refer to the local part so it should be accessible directly. There is no need to abstract too much as one can propose per account/program customization as well (the mail system does it).

**Comment by dtzWill** [comment]
Would it make sense to add a service for this as well?

**Comment by loewenheim** [comment]
> 
> 
> Would it make sense to add a service for this as well?

I definitely think so.

**Comment by teto** [comment]
For sure. You can follow the mbsync split between programs/service so that if someone wants to check his mails manually exclusively he can too.

**Comment by loewenheim** [comment]
I just completed a fairly primitive first version of the calendar infrastructure. This will still need quite a bit of work.

**Comment by teto** [comment]
thanks. will try to have a look in the coming days, if you have an update, now is the time to push :D

File: modules/accounts/calendar.nix [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+#        readOnly = true;
+#        description = ''
+#          Unique identifier of the calendar. This is set to the
+#          attribute name of the calendar configuration.
+#        '';
+#      };
+#    };
+#
+#    config = mkMerge [ { name = name; } ];
+#  };
```
Comment (teto) [inline modules/accounts/calendar.nix]: cleanup

File: modules/accounts/calendar.nix [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+in
+
+{
+  options.accounts.calendar = {
+    basePath = mkOption {
+      type = types.str;
+      default = "${config.home.homeDirectory}/.calendars/";
```
Comment (teto) [inline modules/accounts/calendar.nix]: XDG_DATA_HOME would be a best default IMHO.

File: modules/programs/vdirsyncer-accounts.nix [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+          Show a prompt for the password with the specified
+          text.
+        '';
+      };
+
+      verify = mkOption {
+        type = types.nullOr types.bool;
+        default = null;
+        description = "Verify SSL certificate. Defaults to <literal>true</literal>.";
+      };
```
Comment (teto) [inline modules/programs/vdirsyncer-accounts.nix]: defaults is wrong

File: modules/programs/vdirsyncer-accounts.nix [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+        description = "The type of the storage.";
+      };
+
+      path = mkOption {
+        type = types.nullOr types.str;
+        default = null;
+        description = "The path of the storage.";
+      };
+
+      fileExt = mkOption {
```
Comment (teto) [inline modules/programs/vdirsyncer-accounts.nix]: would be nice to be more precise/give some example. It doesn't seem necessary as I don't have one configured.

File: modules/programs/vdirsyncer-accounts.nix [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+          <link xlink:href="https://vdirsyncer.pimutils.org/en/stable/ssl-tutorial.html#ssl-tutorial"/>
+          for more information.
+        '';
+      };
+
+      auth = mkOption {
+        type = types.nullOr (types.enum ["basic" "digest" "guess"]);
+        default = null;
+        description = ''
+          Authentication settings. The default is <literal>"basic"</literal>.
```
Comment (teto) [inline modules/programs/vdirsyncer-accounts.nix]: maybe rephrase to "vdirsyncer defaults to basic"

**Comment by teto** [comment]
could you share an example of how you configure it please ? to ease bootstrap.

**Comment by loewenheim** [comment]
Sorry for the long delay. This is what my configuration looks like (modulo private information):

```nix
programs.khal.enable = true;
programs.vdirsyncer.enable = true;
programs.vdirsyncer.package = pkgs.vdirsyncerStable;

accounts.calendar = {
  basePath = "$HOME/.calendars";  
  accounts.main = {
    khal.enable = true;

    vdirsyncer = {
      enable = true;
      collections = null;
      metadata = ["color" "displayname"];
      local = {
        type = "filesystem";
        fileExt = ".ics";
      };

      remote = {
        type = "caldav";
        url = "https://dav.mailbox.org/caldav/<some hash>";
        userName = "<my email address>";
        passwordCommand = ["~/.dotfiles/scripts/password.sh" "eMail/mailbox.org"];
      };      
    };
  };
};

accounts.contact = {
  basePath = "$HOME/.contacts";
  accounts.main = {
    vdirsyncer.enable = true;
    vdirsyncer.local = {
      type = "filesystem";
      fileExt = ".vcf";
    };

    vdirsyncer.remote = {
      type = "carddav";
      url = "https://dav.mailbox.org/carddav/<some hash>";
      userName = "<my email address>";
      passwordCommand = ["~/.dotfiles/scripts/password.sh" "eMail/mailbox.org"];
    };
  };
};
```

Is this what you had in mind?

**Comment by teto** [comment]
exactly thanks ! My server is down so I can't exactly test it yet but :
- would be nice to allow a command to fetch the username, like for password (it's possible : http://vdirsyncer.pimutils.org/en/stable/keyring.html)
- rather than having password/passwordCommand, maybe a different behavior wether it's a list or a string ?

**Comment by loewenheim** [comment]
>* would be nice to allow a command to fetch the username, like for password (it's possible : http://vdirsyncer.pimutils.org/en/stable/keyring.html)

There is a `userNameCommand` option, if that’s what you mean.
 
>* rather than having password/passwordCommand, maybe a different behavior wether it's a list or a string ?

The option would also have to cover `passwordPrompt`, which has the same syntax as `passwordCommand`. How would we diffentiate between the two?

File: modules/programs/vdirsyncer-accounts.nix [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+        description = ''
+          Show a prompt for the password with the specified
+          text.
+        '';
+      };
+
+      verify = mkOption {
+        type = types.nullOr types.bool;
+        default = null;
+        description = "Verify SSL certificate. Defaults to <literal>true</literal>.";
```
Comment (rycee) [inline modules/programs/vdirsyncer-accounts.nix]: Like @teto pointed out, the "Defaults to <literal>true</literal>." is wrong and could be removed.

File: modules/accounts/contacts.nix [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+      ]);
+      default = {};
+      description = "List of contacts.";
+    };
+  };
+  config = mkIf (cfg.accounts != {}) {
+    assertions =
+      map (a:
+          {
+            assertion = a.khal.type == "birthdays";
```
Comment (rycee) [inline modules/accounts/contacts.nix]: It seems a bit inappropriate to have an assertion related to khal inside this general file. Perhaps move it to the khal module?

File: modules/accounts/contacts.nix [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+        type = types.str;
+        default = "${cfg.basePath}/${name}";
+        description = "The path of the storage.";
+      };
+    };
+
+    config = mkMerge [
+      {
+        name = name;
+        khal.type = mkOptionDefault "birthdays";
```
Comment (rycee) [inline modules/accounts/contacts.nix]: Could be moved to the `khal-accounts.nix` file?

File: modules/accounts/calendar.nix [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+            assertion = length primaries <= 1;
+            message =
+              "Must have exactly one or zero primary calendar accounts but found "
+              + toString (length primaries)
+              + ", namely "
+              + concatStringsSep ", " primaries;
+          }
+      )
+    ] ++
+      map (a:
```
Comment (rycee) [inline modules/accounts/calendar.nix]: Could be moved to the `khal.nix` file?

File: modules/programs/khal-accounts.nix [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+      type = types.str;
+      default = "*";
+      description = ''
+        The glob expansation to be searched for events or birthdays when type
+        is set to discover.
+      '';
+    };
+  };
+
+  config = mkIf config.khal.enable {
```
Comment (rycee) [inline modules/programs/khal-accounts.nix]: Could be removed.

File: modules/accounts/calendar.nix [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+        default = false;
+        description = ''
+          Whether this is the primary account. Only one account may be
+          set as primary.
+        '';
+      };
+
+      primaryCollection = mkOption {
+        type = types.str;
+        default = if (config.vdirsyncer.collections == null || config.vdirsyncer.collections == []) then
```
Comment (rycee) [inline modules/accounts/calendar.nix]: Ideally defaults in this general module should not be defined in terms of values people set in more specific modules. I would suggest having no default.

File: modules/accounts/calendar.nix [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+      name = mkOption {
+        type = types.str;
+        readOnly = true;
+        description = ''
+          Unique identifier of the calendar. This is set to the
+          attribute name of the calendar configuration.
+        '';
+      };
+
+      path = mkOption {
```
Comment (rycee) [inline modules/accounts/calendar.nix]: This should go into some form of submodule corresponding to the `maildir` submodule in the email account module. Not sure whether there is any widely supported local storage of calendars and contacts. I guess single vcf, vcs, or ics containing multiple entries is common? Or basic directory structures of single files?

Could then have `vcfFile` and `vcfDirectory` as sub-modules. Should be `null`-able for people not wanting to sync locally.

File: modules/accounts/calendar.nix [hunk 15]
```diff
@@ -0,0 +1,111 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+
+  cfg = config.accounts.calendar;
+
+  calendarOpts = { name, config, ... }: {
+    options = {
```
Comment (rycee) [inline modules/accounts/calendar.nix]: I would have expected a caldav sub-module here with some connection information. I saw that `vdirsyncer-accounts.nix` contains that sort of stuff. So would be nice to move that here.

Similarly the contacts module would benefit from having a carddav sub-module.

Comment (loewenheim) [reply modules/accounts/calendar.nix]: What would you obtain these respective submodules to contain?

**Comment by teto** [comment]
I fixed my server. Looking forward to your update :p

**Comment by loewenheim** [comment]
Hm, I’m wondering what the module structure should look like in the light of @rycee’s comment:

>This should go into some form of submodule corresponding to the maildir submodule in the email account module. Not sure whether there is any widely supported local storage of calendars and contacts. I guess single vcf, vcs, or ics containing multiple entries is common? Or basic directory structures of single files?
Could then have vcfFile and vcfDirectory as sub-modules. Should be null-able for people not wanting to sync locally.

What options should a calendar have? I’m going to focus on the local options here; similar questions arise in the remote case. One way to do it would be to have a `local` submodule that contains 

- options such as `path`, `type` (single file or multiple files) and `encoding`;
- a submodule `vdirsyncer` for local configuration that is specific to vdirsyncer.

On the other hand, there could be a `local` submodule that contains only the common configuration and then a `vdirsyncer` submodule that contains everything vdiirsyncer related, both local and remote.

File: modules/accounts/contacts.nix (hunk 10 above)
Comment (loewenheim) [reply modules/accounts/contacts.nix]: The problem here is that `type` logically shouldn’t even be an option for `contact` accounts because `"birthdays"` is the only valid value. To be honest, I don’t know what to do about this; splitting `khal-accounts` in two for calendars and contacts seems cumbersome..

File: modules/accounts/calendar.nix (hunk 15 above)
Comment (teto) [reply modules/accounts/calendar.nix]: the carddav/caldav could contain url, a way to pass on the password, eventually a type like "nextcloud"/"owncloud"/"google" to automatically adapt some parameters like https://github.com/rycee/home-manager/blob/f83c49baa3cf4de627bcfd3c70730478aaa3998a/modules/accounts/email.nix#L87

**Comment by teto** [comment]
I've commented for the remote submodule in the code.
I agree with a `local` submodule that for now I only know for sure the dest path can be safely abstracted.
 
I would like to see this merged so let's say for now when in doubt that a setting is not abstract enough, assign to vdirsyncer-account. This could be revisited later. The problem is there are not many cardavl/caldav clients to look for (calcurse ?).

**Comment by loewenheim** [comment]
I have restructured the contact, calendar, vdirsyncer and khal modules. The configuration I previously posted would now look like this:

```
programs.khal.enable = true;
programs.vdirsyncer.enable = true;
programs.vdirsyncer.package = pkgs.vdirsyncerStable;

accounts.calendar = {
  accounts.main = {
    khal.enable = true;

    vdirsyncer = {
      enable = true;
      collections = null;
      metadata = ["color" "displayname"];      
    };

    local = {
      type = "filesystem";
      fileExt = ".ics"
    };

    remote = {
      type = "caldav";
      url = "https://dav.mailbox.org/caldav/<some hash>";
      userName = "<my email address>";
      passwordCommand = ["~/.dotfiles/scripts/password.sh" "eMail/mailbox.org"];
    };
  };
};

accounts.contact = {
  basePath = "$HOME/.contacts";
  accounts.main = {
    vdirsyncer.enable = true;

    local = {
      type = "filesystem";
      fileExt = ".vcf";
    };

    remote = {
      type = "carddav";
      url = "https://dav.mailbox.org/carddav/<some hash>";
      userName = "<my email address>";
      passwordCommand = ["~/.dotfiles/scripts/password.sh" "eMail/mailbox.org"];
    };
  };
};
```

One thing I’m not happy with is that I split `khal-accounts.nix` up into `khal-accounts.nix` and `khal-calendar-accounts.nix`. The former contains the options for both calendar and contact accounts, whereas the latter contains the `type` option that only makes sense for calendars. I couldn’t figure out how to do this in one file.

File: modules/services/vdirsyncer.nix [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+  config = mkIf cfg.enable {
+    systemd.user.services.vdirsyncer = {
+      Unit = {
+        Description = "vdirsyncer calendar&contacts synchronization";
+        PartOf = [ "network-online.target" ];
+      };
+
+      Service = {
+        Type = "oneshot";
+        # TODO `vdirsyncer discover`
```
Comment (teto) [inline modules/services/vdirsyncer.nix]: the discover part is annoying. It's not necessary when collections != null though. In the long run, vdirsyncer should be able to do without it but as it has no developers anymore :'(

Comment (teto) [reply modules/services/vdirsyncer.nix]: you could mention it in the `collections` setting description btw.

**Comment by teto** [comment]
Thanks for the changes it looks better.
- when using basePath = "$HOME/calendars"; vdirsyncer would write in the wrong directory since it won't interpret the variable. My manual config had for instance `path.fetch = ["command", "sh", "-c", "echo $XDG_DATA_HOME/calendars/perso"]` . I don't see this as a blocker for now but it's good to know.
- there are trailing spaces a bit everywhere, please clean them up

**Comment by loewenheim** [comment]
I’ve dealt with the trailing spaces.

**Comment by teto** [comment]
Thanks. Works for me. I think it's mergeable.

**Comment by loewenheim** [comment]
Thank you very much. This has been a great learning experience for me!

**Comment by teto** [comment]
@rycee I am tempted to click merge but as it's a huge PR (+1,037 lines) setting some generic infra, I'll let you manage the rest :p

File: modules/programs/vdirsyncer-accounts.nix [hunk 17]
```diff
@@ -0,0 +1,217 @@
...
+      type = types.nullOr types.str;
+      default = null;
+      description = ''
+        The user agent to report to the server.
+        Defaults to <literal>"vdirsyncer"</literal>.
+      '';
+    };
+
+    postHook = mkOption {
+      type = types.nullOr types.str;
```
Comment (rycee) [inline modules/programs/vdirsyncer-accounts.nix]: Could have
```
type = types.lines;
default = "";
```
and skip generating the actual post hook script if `postHook == ""`.

File: modules/programs/vdirsyncer-accounts.nix [hunk 18]
```diff
@@ -0,0 +1,217 @@
...
+    tokenFile = mkOption {
+      type = types.nullOr types.str;
+      default = null;
+      description = ''
+        A file path where access tokens
+        are stored.
+      '';
+    };
+
+    clientId = mkOption {
```
Comment (rycee) [inline modules/programs/vdirsyncer-accounts.nix]: Can remove so that the users are encouraged to use `clientIdCommand` instead.

File: modules/programs/vdirsyncer-accounts.nix [hunk 19]
```diff
@@ -0,0 +1,217 @@
...
+
+        OAuth credentials, obtained from the Google API Manager.</para>
+
+        <para> See
+        <link xlink:href="https://vdirsyncer.pimutils.org/en/stable/config.html#google"/>
+        for more information.
+      '';
+    };
+
+    clientSecret = mkOption {
```
Comment (rycee) [inline modules/programs/vdirsyncer-accounts.nix]: Similarly, can remove so that the users are encouraged to use `clientSecretCommand` instead.

File: modules/programs/vdirsyncer.nix [hunk 20]
```diff
@@ -0,0 +1,269 @@
...
+
+  boolString = b: if b then "true" else "false";
+
+  localStorage = a:
+  filterAttrs (_: v: v != null)
+  ((getAttrs [ "type" "fileExt" "encoding" ] a.local) // {
+    path = a.local.path;
+    postHook = if a.vdirsyncer == null
+      then null
+      else a.vdirsyncer.postHook;
```
Comment (rycee) [inline modules/programs/vdirsyncer.nix]: I'd suggest writing the post hook commands to a shell script, e.g. to have something like
```
postHook =
  pkgs.writeShellScriptBin "post-hook" a.vdirsyncer.postHook
  + "/bin/post-hook";
```
This will allow the use of multiline scripts and the `types.lines` type.

File: modules/modules.nix [hunk 21]
```diff
@@ -123,6 +127,7 @@ let
     (loadModule ./services/tahoe-lafs.nix { })
     (loadModule ./services/udiskie.nix { })
     (loadModule ./services/unclutter.nix { })
+    (loadModule ./services/vdirsyncer.nix { })
```
Comment (rycee) [inline modules/modules.nix]: Since the service uses systemd it is limited to Linux:
```suggestion
    (loadModule ./services/vdirsyncer.nix { condition = hostPlatform.isLinux; })
```

File: modules/accounts/contacts.nix [hunk 22]
```diff
@@ -0,0 +1,157 @@
...
+        type = types.nullOr (types.listOf types.str);
+        default = null;
+        example = [ "pass" "caldav" ];
+        description = ''
+          A command that prints the password to standard
+          output.
+        '';
+      };
+
+      passwordPrompt = mkOption {
```
Comment (rycee) [inline modules/accounts/contacts.nix]: Similarly, perhaps safest to remove.

File: modules/accounts/calendar.nix [hunk 23]
```diff
@@ -0,0 +1,193 @@
...
+        type = types.nullOr (types.listOf types.str);
+        default = null;
+        example = [ "pass" "caldav" ];
+        description = ''
+          A command that prints the password to standard
+          output.
+        '';
+      };
+
+      passwordPrompt = mkOption {
```
Comment (rycee) [inline modules/accounts/calendar.nix]: I suspect this is pretty specific for vdirsyncer, perhaps could remove this as well. Alternatively move to the vdirsyncer-account file?

**Comment by loewenheim** [comment]
Thank you for the kind words, @rycee. I’ve implemented your suggestions.

File: modules/programs/vdirsyncer.nix [hunk 24]
```diff
@@ -0,0 +1,264 @@
...
+        then [ "encoding" ]
+      else if (t == "google_calendar") then
+        [ "timeRange" "itemTypes" "clientIdCommand" "clientSecretCommand" ]
+      else if (t == "google_contacts") then [ "clientIdCommand" "clientSecretCommand" ]
+      else throw "Unrecognized storage type: ${t}";
+
+      assertStorage = n: v:
+      [
+        {
+          assertion = length (filter (x: x) [
```
Comment (rycee) [inline modules/programs/vdirsyncer.nix]: With the options gone, this assertion is no longer needed.

File: modules/accounts/calendar.nix (hunk 4 above)
Comment (rycee) [reply modules/accounts/calendar.nix]: An alternative is to simply leave out the default value and force the user to make an conscious decision.

**Comment by rycee** [comment]
I have pushed a squashed version of this branch to prepare this for merge. See #754.

File: modules/services/vdirsyncer.nix (hunk 16 above)
Comment (pacien) [reply modules/services/vdirsyncer.nix]: What about running `vdirsyncer metasync` in addition to `sync` to have calendar colours and some other metadata synchronised too?

File: modules/programs/vdirsyncer-accounts.nix [hunk 25]
```diff
@@ -0,0 +1,193 @@
...
+      '';
+    };
+
+    verify = mkOption {
+      type = types.nullOr types.bool;
+      default = null;
+      description = "Verify SSL certificate.";
+    };
+
+    verifyFingerprint = mkOption {
```
Comment (pacien) [inline modules/programs/vdirsyncer-accounts.nix]: `verify` and `verify_fingerprint` have been removed and replaced by `verify_cert` in recent versions of vdirsyncer, including the one packaged on NixOS 19.03.

See https://github.com/pimutils/vdirsyncer/issues/780, [latest/ssl-tutorial](http://vdirsyncer.pimutils.org/en/latest/ssl-tutorial.html) vs [stable/ssl-tutorial](http://vdirsyncer.pimutils.org/en/stable/ssl-tutorial.html).

**Comment by jorsn** [comment]
@loewenheim and the others: Are you still interested in this? I'd very much like to get #754 merged. However, in the last half year, I seem to be the only one working (a bit) on this, at least in public.

**Comment by loewenheim** [comment]
I’m sorry, I totally lost sight of this and don’t know when I could get back to it. I would appreciate someone else taking over.

**Comment by DamienCassou** [comment]
I'm interested. What needs to be done?
